torch/csrc/Module.cpp

```diff
@@ -570,6 +573,118 @@ PyObject *THPModule_fromDLPack(PyObject *_unused, PyObject *data)
   return torch::createPyObject(atensor);
 }
 
+// In cases like data loader, if a worker process die due to bus error/segfault
+// or just hang, the main process, if implemented with multiprocessing.queue.,
+// will hang waiting for data. This is difficult to avoid on PyTorch side as it
+// can be caused by limited shm, other libraries users call in the workers. The
+// following methods is an effort to do our best provide some error message to
+// users when such unfortunate events happen.
+
+// Critical signal handlers should be registered on worker processes before
+// doing work.
+// Python handle is _set_worker_signal_handlers().
+#define SIGNAL_HANDLER(SIGNAL, HANDLER_NAME, ERROR_MSG)                       \
+static void HANDLER_NAME(int sig)                                             \
+{                                                                             \
+    write(fileno(stderr), ERROR_MSG, strlen(ERROR_MSG));                      \
+    _exit(EXIT_FAILURE);                                                      \
+}
+
+SIGNAL_HANDLER(SIGBUS, handler_SIGBUS, "ERROR: Unexpected bus error encountered in worker. "
+  "This might be caused by insufficient shared memory (shm).\n");
+SIGNAL_HANDLER(SIGSEGV, handler_SIGSEGV, "ERROR: Unexpected segmentation fault encountered in worker.\n");
+
+static std::mutex worker_pid_mutex;
+static std::vector<pid_t> worker_pid_vec = {};
+// The following are needed since std::vector is not asynchronous safe.
+static pid_t *worker_pids;
+static size_t num_worker_pids = 0;
+
+PyObject *THPModule_setWorkerSignalHandlers(PyObject *module, PyObject *arg) {
```

Comment:
Let me know if I should move these methods to a separate file.

Reply:
Yeah, I think it'd be better in a separate file. In the new file's header you can export a null-terminated `PyMethodDef[]`. In Module.cpp use THPUtils_addPyMethodDefs to add it to the methods vector.